File: users-service/src/main/java/net/plshark/users/repo/springdata/SpringDataUserGroupsRepository.java
package net.plshark.users.repo.springdata;

import java.util.Objects;
import net.plshark.users.model.Group;
import net.plshark.users.repo.UserGroupsRepository;
import org.springframework.data.r2dbc.core.DatabaseClient;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/**
 * User groups repository using spring data
 */
@Repository
public class SpringDataUserGroupsRepository implements UserGroupsRepository {

    private final DatabaseClient client;

    public SpringDataUserGroupsRepository(DatabaseClient client) {
        this.client = Objects.requireNonNull(client);
    }

    @Override
    public Flux<Group> getGroupsForUser(long userId) {
        return client.execute()
                .sql("SELECT * FROM groups g INNER JOIN user_groups ug ON g.id = ug.group_id WHERE ug.user_id = :id")
                .bind("id", userId)
                .map(SpringDataGroupsRepository::mapRow)
                .all();
    }

    @Override
    public Mono<Void> insert(long userId, long groupId) {
        return client.execute()
                .sql("INSERT INTO user_groups (user_id, group_id) values (:userId, :groupId)")
                .bind("userId", userId)
                .bind("groupId", groupId)
                .then();
    }

    @Override
    public Mono<Void> delete(long userId, long groupId) {
        return client.execute()
                .sql("DELETE FROM user_groups WHERE user_id = :userId AND group_id = :groupId")
                .bind("userId", userId)
                .bind("groupId", groupId)
                .then();
    }

    @Override
    public Mono<Void> deleteUserGroupsForUser(long userId) {
        return client.execute()
                .sql("DELETE FROM user_groups WHERE user_id = :userId")
                .bind("userId", userId)
                .then();
    }

    @Override
    public Mono<Void> deleteUserGroupsForGroup(long groupId) {
        return client.execute()
                .sql("DELETE FROM user_groups WHERE group_id = :groupId")
                .bind("groupId", groupId)
                .then();
    }
}

File: user-error-server/src/main/kotlin/net/plshark/users/repo/springdata/SpringDataUserGroupsRepository.kt
package net.plshark.users.repo.springdata

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.reactive.asFlow
import net.plshark.users.model.Group
import net.plshark.users.model.Role
import net.plshark.users.repo.UserGroupsRepository
import org.springframework.r2dbc.core.DatabaseClient
import org.springframework.r2dbc.core.await
import org.springframework.stereotype.Repository

/**
 * User groups repository using spring data
 */
@Repository
class SpringDataUserGroupsRepository(private val client: DatabaseClient) : UserGroupsRepository {

    override fun findGroupsByUserId(userId: Long): Flow<Group> {
        val sql = "SELECT * FROM groups g INNER JOIN user_groups ug ON g.id = ug.group_id WHERE ug.user_id = :id"
        return client.sql(sql)
            .bind("id", userId)
            .map { row -> SpringDataGroupsRepository.mapRow(row) }
            .all()
            .asFlow()
    }

    override fun findGroupRolesByUserId(userId: Long): Flow<Role> {
        val sql = "SELECT r.* from roles r, user_groups ug, group_roles gr WHERE ug.user_id = :userId AND " +
            "gr.group_id = ug.group_id AND r.id = gr.role_id"
        return client.sql(sql)
            .bind("userId", userId)
            .map { row -> SpringDataRolesRepository.mapRow(row) }
            .all()
            .asFlow()
    }

    override suspend fun insert(userId: Long, groupId: Long) {
        return client.sql("INSERT INTO user_groups (user_id, group_id) values (:userId, :groupId)")
            .bind("userId", userId)
            .bind("groupId", groupId)
            .await()
    }

    override suspend fun deleteById(userId: Long, groupId: Long) {
        return client.sql("DELETE FROM user_groups WHERE user_id = :userId AND group_id = :groupId")
            .bind("userId", userId)
            .bind("groupId", groupId)
            .await()
    }
}
